### ISSUE TITLE
Remove `TriggerHardForkAtEpoch`

### ISSUE THREAD

**Issue by amesgen** [input-output-hk/ouroboros-consensus#416]
Due to #395, we have good reason to use case is **fundamentally**/**crucially** relying on `TriggerHardForkAtEpoch`. Hence, we would like to remove them to remove them to simplify the possible trajectories of the HFC (also see [this HWW section](https://input-output-hk.github.io/ouroboros-consensus/docs/for-developers/HardWonWisdom#the-safe-zone-semantics)).

Goals of this ticket:

 - Remove the `TriggerHardForkAtEpoch` constructor, and adapt documentation accordingly.
 - Introduce a different way for users to specify that they wish to directly skip to a later era. (The current interface/configuration options could be kept as is for a migration period.)
 - Provide guidance to migrate away from the remaining use cases of `TriggerHardForkAtEpoch` mentioned in #395.

**Comment by amesgen** [comment]
Today, we discussed the following idea that should supersed supersede `TriggerXXXHardForkAtEpoch` (where `XXX` is an era): Introducing configuration that would cause the ledger to increase the (major) protocol version to a specific value.

For example, the node `.yaml` file could contain configuration like
```yaml
ProtocolVersionUpdates:
- epoch: 5
  protocolVersion: 10
```
and the Ledger would hence update the major protocol version to 10 in epoch 5[^1].

This way, "scripted" era transitions on a testnet (e.g. Babbage at protocol version `8` for the first two epochs, then Conway at protocol version `9` for two epochs, then Conway at protocol version `10` for Chang+1 which unlocks new Plutus functionalities) are easily possible without actually having to submit any governance-related transactions. In particular, intra-era HFs (like to Chang+1 in the example) are not possible with the current `TriggerHardForkAtEpoch` mechanism.

The downside of this approach is that it makes Consensus/the HFC simpler at the expense of making Ledger more complex, so this definitely requires further thought/discussion whether this pays its weight.

[^1]: This requires a bit of thought re what should happen if the protocol version is already larger than 10, or whether this update should take precedence over a regular governance-induced update to the protocol version.

**Comment by nfrisby** [comment]
I'd summarize the state of https://github.com/IntersectMBO/ouroboros-consensus/issues/395 and this 416 Issue as follows.

1. See Issue 395 for reports of how this version is being used: mostly to skip eras in epoch 0, but also a few scheduled transitions in epoch 1 or 2 eg.

2. The Slack thread https://input-output-rnd.slack.com/archives/CFKLUH4R0/p1731494758341649 additionally demonstrates an example where someone was confused that the `TriggerHardForkAtEpoch` flags in the node configuration caused an era transition but _did not_ increment the major protocol version. Especially due to the logic of intra-era hard forks branching on the protocol version itself, this strongly indicates to me that it's entirely non-sensical for the HFC to break the synchronization of the ledger's era with the protocol version protocol parameter.

3. So I see three options.

- The HFC mucks with the ledger's state to update the protocol version parameter when transitioning due to a given `TriggerHardForkAtEpoch` configuration. (I do not like this; it's upside down.)
- We upstream the `TriggerHardForkAtEpoch` interpretation into the Ledger itself instead of the HFC. (I would like this, but Ledger will probably hesitate.) (This is the previous comment on this 416 Issue.)
- The downstream users that want to schedule transitions at certain epochs no longer get to do so via configuration data and must instead submit the necessary governance transactions at the necessary time. (I can't imagine the downstream users will like this, but perhaps Ledger and Consensus can try to make the necessary preparations as painless as possible.) (IE the description of this 416 Issue itself.)

**Comment by nfrisby** [comment]
I opened https://github.com/IntersectMBO/cardano-ledger/issues/4898 to prompt the Ledger to weigh-in on the above comment ^^^.

**Comment by nfrisby** [comment]
New plan: I'll draft a CIP with migration instructions.

**Comment by nfrisby** [comment]
Prior to a CIP, I'm writing it up for "internal" users (who are plausibly the only users) here https://github.com/IntersectMBO/ouroboros-consensus/pull/1408

While doing so, I finally understood enough of the detailed dynamics to realize that this task ...

> Introduce a different way for users to specify that they wish to directly skip to a later era. (The current interface/configuration options could be kept as is for a migration period.)

... is merely a flag indicating whether to skip Byron.

For the Cardano ledger, at least, the [table maintained alongside CIP-0059](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0059/feature-table.md) is _not_ optional. The Ledger rules themselves assume that eg Alonzo can only run with version 5 or 6.

- If `SkipByron` is absent or not set, then the Shelley Genesis file must contain `protocolVersion: {major: 2, minor: 0}` and the Cardano node should initialize in Byron with protocol version 0.
- If `SkipByron` is set, then the Shelley Genesis file's protocol version must be >=2.0, and the Cardano node should skip to the corresponding era at the onset of slot 0, as determined by the CIP-0059 table.

**Comment by nfrisby** [comment]
I set this to status "Help needed" in the Consensus Team Backlog, since we're now blocked on those downstream Issues.
